drivers/QSPI.h

```diff
@@ -0,0 +1,232 @@
+ *
+ * The default format is set to Quad-SPI(4-4-4), and a clock frequency of 1MHz
+ * Most QSPI devices will also require Chip Select which is indicated by ssel. 
+ *
+ * @note Synchronization level: Thread safe
+ *
+ * Example:
+ * @code
+ * // Write 4 byte array to a QSPI slave, and read the response, note that each device will have its specific read/write/alt values defined
+ *
+ * #include "mbed.h"
+ *
+ * // hardware ssel (where applicable)
+ * QSPI qspi_device(p5, p6, p7, p8, p9, p10); // io0, io1, io2, io3, sclk, ssel
+ *
+ *
+ * int main() {
+ *     char tx_buf[] = { 0x11, 0x22, 0x33, 0x44 };    
+ *     char rx_buf[4];    
+ *     int buf_len = sizeof(tx_buf); 
+ *     
+ *     int result = qspi_device.write( 0x12 , 0x100000 , 0 , tx_buf, &buf_len );
+ *     if( !result ) printf("Write failed");
+ *     int result = qspi_device.read( 0x13 , 0x100000 , 0 , rx_buf, &buf_len );
+ *     if( !result ) printf("Read failed");
+ *
+ * }
+ * @endcode
+ * @ingroup drivers
+ */
+class QSPI : private NonCopyable<QSPI> {
+
+public:
+
+    /** Create a QSPI master connected to the specified pins
+     *
+     *  io0-io3 is used to specify the Pins used for Quad SPI mode
+     *
+     *  @param io0-io3 IO pins used for sending/receiving data during data phase of a transaction
+     *  @param sclk QSPI Clock pin
+     *  @param ssel QSPI chip select pin
+     */
+    QSPI(PinName io0, PinName io1, PinName io2, PinName io3, PinName sclk, PinName ssel=NC);
+
+    /** Configure the data transmission format
+     *
+     *  @param inst_width Bus width used by instruction phase(Valid values are 1,2,4)
+     *  @param inst_size Size in bits used by instruction phase(Valid values are NONE,8,16,24,32)
+     *  @param address_width Bus width used by address phase(Valid values are 1,2,4)
+     *  @param address_size Size in bits used by address phase(Valid values are NONE,8,16,24,32)
+     *  @param alt_width Bus width used by alt phase(Valid values are 1,2,4)
+     *  @param alt_size Size in bits used by alt phase(Valid values are NONE,8,16,24,32)
+     *  @param data_width Bus width used by data phase(Valid values are 1,2,4)
+     *  @param dummy_cycles Number of dummy clock cycles to be used after alt phase
+     *
+     * @endcode
+     */
+    bool configure_format(int inst_width = QSPI_DEFAULT_INST_WIDTH, 
+                   int address_width = QSPI_DEFAULT_ADDRESS_WIDTH, 
+                   int address_size = QSPI_DEFAULT_ADDRESS_SIZE,
+                   int alt_width = QSPI_DEFAULT_ALT_WIDTH, 
+                   int alt_size = QSPI_DEFAULT_ALT_SIZE,   
+                   int data_width = QSPI_DEFAULT_DATA_WIDTH,                             
+                   int dummy_cycles = QSPI_DEFAULT_DUMMY_CYCLES,
+                   int mode = 0);
+
+    /** Initialize QSPI interface
+     *
+     *  This function must be called before doing any operation on the QSPI bus to initialize the interface
+     */
+    bool initialize();
```

Comment:
This is to enhance clarity and flexibility in QSPI driver usage. Also, since we are providing configure format function we don't want to initialize the HAL before the APP has set the communication format. This also makes it easier for APPs to re-configure and re-initialize using the same object without having to use different objects if need be.

Reply:
Not convinced here.  we are providing defaults in the most arguments there, therefore using these defaults we can set-up the bus in ctor and object should be ready. no more init(). This init does not even have arguments.

````
QSPIclass qspi;
qspi.write(command, data); // using default values

// I assume often a user reconfigure to their needs
qspi.freq();
qspi.set_somethingelse();
// bus ready
qspi.write(); //
````